### ISSUE TITLE
cKDTree not indexing properly for arrays with too many elements

### ISSUE THREAD

**Issue by sheerbordom** [scipy/scipy#5077]
Please forgive my limited understanding of `cKDTree`--I will articulate the issue as clearly as I can.

I've been having a problem using `cKDTree.sparse_distance_matrix` on large arrays of points. When my array contains ten points or less, `cKDTree` accurately indexes the nearest neighbors for every point. Using `cKDTree.sparse_distance_matrix`, then, produces an array with the distance between each pair of neighboring points AND the reciprocal pair (e.g. the distance from (1, 1) to (2, 2) AND the identical distance from (2, 2) to (1, 1) are both indexed). Thus, the distance matrix is the same as its transpose.

Input: 
```
import numpy as np
from scipy.spatial import cKDTree

#Randomly generated array of points for demonstration
just_fine = np.array(np.random.randn(10, 2), dtype = int) 

tree = cKDTree(just_fine)
D1 = tree.sparse_distance_matrix(tree, 3)

print np.array(D1.todense(), dtype = int)
print np.array(D1.todense().T, dtype = int)

#No difference between D1 and its transpose
print np.array(D1.todense() - D1.todense().T, dtype = int)
```
Output:
```
[[0 0 2 2 2 2 2 2 2 2]
 [0 0 1 1 1 1 1 1 1 1]
 [2 1 0 0 0 2 1 0 0 0]
 [2 1 0 0 0 2 1 0 0 0]
 [2 1 0 0 0 2 1 0 0 0]
 [2 1 2 2 2 0 1 2 2 2]
 [2 1 1 1 1 1 0 1 1 1]
 [2 1 0 0 0 2 1 0 0 0]
 [2 1 0 0 0 2 1 0 0 0]
 [2 1 0 0 0 2 1 0 0 0]]

[[0 0 2 2 2 2 2 2 2 2]
 [0 0 1 1 1 1 1 1 1 1]
 [2 1 0 0 0 2 1 0 0 0]
 [2 1 0 0 0 2 1 0 0 0]
 [2 1 0 0 0 2 1 0 0 0]
 [2 1 2 2 2 0 1 2 2 2]
 [2 1 1 1 1 1 0 1 1 1]
 [2 1 0 0 0 2 1 0 0 0]
 [2 1 0 0 0 2 1 0 0 0]
 [2 1 0 0 0 2 1 0 0 0]]

[[0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]]
```

However, when I augment the array so that it contains 11 or more points, reciprocal pairs are NOT always indexed. A distance matrix for the set of points, therefore, is NOT the same as its transpose.

Input:  
```
too_many = np.array(np.random.randn(11, 2), dtype = int)

tree = cKDTree(too_many)
D2 = tree.sparse_distance_matrix(tree, 3)

print np.array(D2.todense(), dtype = int)
print np.array(D2.todense().T, dtype = int)

#Highlighting discrepancy between D2 and its transpose
print np.array(D2.todense() - D2.todense().T, dtype = int)
```
Output:
```
[[0 1 2 0 0 1 0 0 0 0 0]
 [1 0 2 1 0 1 1 0 1 1 1]
 [2 2 0 2 0 3 2 0 2 2 2]
 [0 1 2 0 0 1 0 0 0 0 0]
 [2 0 0 2 0 2 2 1 2 2 2]
 [1 1 3 1 0 0 1 0 1 1 1]
 [0 1 2 0 0 1 0 0 0 0 0]
 [1 2 2 1 1 1 1 0 1 1 1]
 [0 1 2 0 0 1 0 0 0 0 0]
 [0 1 2 0 0 1 0 0 0 0 0]
 [0 1 2 0 0 1 0 0 0 0 0]]

[[0 1 2 0 2 1 0 1 0 0 0]
 [1 0 2 1 0 1 1 2 1 1 1]
 [2 2 0 2 0 3 2 2 2 2 2]
 [0 1 2 0 2 1 0 1 0 0 0]
 [0 0 0 0 0 0 0 1 0 0 0]
 [1 1 3 1 2 0 1 1 1 1 1]
 [0 1 2 0 2 1 0 1 0 0 0]
 [0 0 0 0 1 0 0 0 0 0 0]
 [0 1 2 0 2 1 0 1 0 0 0]
 [0 1 2 0 2 1 0 1 0 0 0]
 [0 1 2 0 2 1 0 1 0 0 0]]

[[ 0  0  0  0 -2  0  0 -1  0  0  0]
 [ 0  0  0  0  0  0  0 -2  0  0  0]
 [ 0  0  0  0  0  0  0 -2  0  0  0]
 [ 0  0  0  0 -2  0  0 -1  0  0  0]
 [ 2  0  0  2  0  2  2  0  2  2  2]
 [ 0  0  0  0 -2  0  0 -1  0  0  0]
 [ 0  0  0  0 -2  0  0 -1  0  0  0]
 [ 1  2  2  1  0  1  1  0  1  1  1]
 [ 0  0  0  0 -2  0  0 -1  0  0  0]
 [ 0  0  0  0 -2  0  0 -1  0  0  0]
 [ 0  0  0  0 -2  0  0 -1  0  0  0]]
```

I'm attributing the discrepancy to a bug in `cKDTree`. There doesn't seem to be any clear pattern to which reciprocal pairs are omitted in the indexing. The error does not occur when I substitute `KDTree` for `cKDTree`.

**Comment by WarrenWeckesser** [comment]
I can reproduce this with scipy 0.15.1, but it works fine with a build of a recent copy of master.  Does anyone recognize this problem, and perhaps know which change fixed it?

**Comment by rgommers** [comment]
Not sure exactly, but probably fixed by one of @sturlamolden's PRs to rewrite ``cKDTree`` in C++.

**Comment by rgommers** [comment]
Indeed fixed in master, so closing.

**Comment by sturlamolden** [comment]
I guess we should add a regression test for this to the test suite so it does not accidentally show up again.

**Comment by sturlamolden** [comment]
I am not sure why it appears fixed in master, I did not do anything to correct it, but **please reopen**. There is a severe logic error in Pat Varilly's Cython code causing this. I have not noticed it before now. It uses the logic from ```count_pairs```, but it should not because ```sparse_distance_matrix``` works on two separate kd-trees. @aarchiba 's code in KDTree is the correct one. Compare with that. I will squash this bug promptly. Please tag it as a critical defect.

**Comment by sturlamolden** [comment]
ping @rgommers

**Comment by WarrenWeckesser** [comment]
@sturlamolden: I haven't looked into the bug, but I'll take your word for it that the issue should be reopened.

**Comment by sturlamolden** [comment]
It is an easy fix BTW. I will post a PR later today.

**Comment by sturlamolden** [comment]
The bug is still in master but harder to tickle. You should be able to reproduce it if you construct the ```cKDTree``` object with keyword arguments ```balanced_tree=False``` and ```compact_nodes=False```.

**Comment by sturlamolden** [comment]
Oh yes, and I know where it it, why it happens and what to do about it, so it goes away very soon :)

**Comment by sturlamolden** [comment]
And it went below the test suite's radar because it only happens when ```sparse_distance_matrix``` is called with the same ```cKDTree``` instance as argument.

**Comment by sturlamolden** [comment]
This proves it is still present in master.

```Python
def test_against_logic_error_regression():
    # regression test for gh-5077 logic error
    np.random.seed(0)
    too_many = np.array(np.random.randn(18, 2), dtype=int)
    tree = cKDTree(too_many, balanced_tree=False, compact_nodes=False)
    d = tree.sparse_distance_matrix(tree, 3).todense()
    assert_array_almost_equal(d, d.T, decimal=14)
```

**Comment by sturlamolden** [comment]
This code proves that the problem is still i master but fixed in the PR (added to the test suite).

```Python
np.random.seed(0)
too_many = np.array(np.random.randn(18, 2), dtype=int)
tree = cKDTree(too_many, balanced_tree=False, compact_nodes=False)
d = tree.sparse_distance_matrix(tree, 3).todense()
assert_array_almost_equal(d, d.T, decimal=14)
```

**Comment by sturlamolden** [comment]
This issue can be closed when #5088 is pulled to master and 0.16.x.

**Comment by rgommers** [comment]
fixed in gh-5088, closing
